### PR TITLE
Rewrite statistics during OPTIMIZE in Iceberg

### PR DESCRIPTION
**Pull request by pajaks**
## Description
Fixes https://github.com/trinodb/trino/issues/19992

## Release notes

(x) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
( ) Release notes are required, with the following suggested text:

### CODE REVIEW

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java [hunk 1]
```diff
@@ -5162,7 +5162,7 @@ public void testOptimizeCleansUpDeleteFiles()
         assertThat(allDataFilesAfterDelete).hasSize(6);
 
         // For optimize we need to set task_min_writer_count to 1, otherwise it will create more than one file.
-        computeActual(withSingleWriterPerTask(getSession()), "ALTER TABLE " + tableName + " EXECUTE OPTIMIZE WHERE regionkey = 4");
+        computeActual(withSingleWriterPerTask(getSession()), "ALTER TABLE " + tableName + " EXECUTE OPTIMIZE WHERE regionkey = 3");
```
Comment (pajaks) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java:5165]: Row deleted has `regionkey = 3`

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergStatistics.java [hunk 2]
```diff
@@ -982,6 +982,70 @@ public void testShowStatsAfterExpiration()
...
+
+        assertQuery(
+                "SHOW STATS FOR " + tableName,
+                """
+                        VALUES
+                          ('key', null, 4, 0, null, '1', '4'), -- NDV present
+                          (null,  null, null, null, 4, null, null)""");
+
+        assertUpdate("DROP TABLE " + tableName);
+    }
```
Comment (alexjo2144) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergStatistics.java:1048]: Nit: newline between the tests
```suggestion
    }

```

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TableStatisticsWriter.java [hunk 3]
```diff
@@ -112,6 +108,23 @@ public StatisticsFile writeStatisticsFile(
...
+
+    public StatisticsFile rewriteStatisticsFile(ConnectorSession session, Table table, long snapshotId)
+    {
+        TableOperations operations = ((HasTableOperations) table).operations();
+        FileIO fileIO = operations.io();
+        // This will rewrite old statistics file as ndvSketches map is empty
+        return writeStatisticsFile(session, table, fileIO, snapshotId, Map.of());
+    }
+
+    private GenericStatisticsFile writeStatisticsFile(ConnectorSession session, Table table, FileIO fileIO, long snapshotId, Map<Integer, CompactSketch> ndvSketches)
```
Comment (alexjo2144) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TableStatisticsWriter.java:122]: Do we need to copy the files to a new copy, or can we reference the existing files with the new snapshot in-place?

Comment (pajaks) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TableStatisticsWriter.java:122]: Referencing the same file from two snapshots does not work. File will be deleted due to cleanup of expired snapshot.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java [hunk 4]
```diff
@@ -1494,6 +1495,29 @@ private void finishOptimize(ConnectorSession session, IcebergTableExecuteHandle
...
+        long newSnapshotId = transaction.table().currentSnapshot().snapshotId();
+        transaction = null;
+
+        // TODO (https://github.com/trinodb/trino/issues/15439): it would be good to publish data and stats atomically
+        beforeWriteSnapshotId.ifPresent(previous ->
+                verify(previous != newSnapshotId, "Failed to get new snapshot ID "));
+
+        try {
+            beginTransaction(catalog.loadTable(session, executeHandle.getSchemaTableName()));
+            StatisticsFile newStatsFile = tableStatisticsWriter.rewriteStatisticsFile(session, icebergTable, newSnapshotId);
```
Comment (findinpath) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:1509]: Why don't you reuse the output from one line above?

```
catalog.loadTable(session, executeHandle.getSchemaTableName())
```

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java [hunk 5]
```diff
@@ -1494,6 +1495,29 @@ private void finishOptimize(ConnectorSession session, IcebergTableExecuteHandle
         rewriteFiles.validateFromSnapshot(snapshot.snapshotId());
         commit(rewriteFiles, session);
         transaction.commitTransaction();
+
+        // TODO (https://github.com/trinodb/trino/issues/15439) this may not exactly be the snapshot we committed, if there is another writer
+        long newSnapshotId = transaction.table().currentSnapshot().snapshotId();
+        transaction = null;
+
+        // TODO (https://github.com/trinodb/trino/issues/15439): it would be good to publish data and stats atomically
+        beforeWriteSnapshotId.ifPresent(previous ->
```
[the remark this answers is not in the document]
Comment (pajaks) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:1504]: I think `rollback_to_snapshot` can cause that is some special case and race condition. I followed what is done for finishInsert where we also write statistics file.
https://github.com/trinodb/trino/blob/23569fcb87ba9e28926754afc068d151c0084b14/plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java#L1132-L1134

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java [hunk 6]
```diff
@@ -1434,6 +1434,7 @@ private void finishOptimize(ConnectorSession session, IcebergTableExecuteHandle
     {
         IcebergOptimizeHandle optimizeHandle = (IcebergOptimizeHandle) executeHandle.getProcedureHandle();
         Table icebergTable = transaction.table();
+        Optional<Long> beforeWriteSnapshotId = Optional.ofNullable(icebergTable.currentSnapshot()).map(Snapshot::snapshotId);
```
[the remark this answers is not in the document]
Comment (pajaks) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java]: When table is empty currentSnapshot can be null, we have PT `testAlterTableExecuteProceduresOnEmptyTable` for that which fails with `NullPointerException`.

**Review by findinpath — APPROVED** [review]

**Review by alexjo2144 — APPROVED** [review]

**Review by homar — APPROVED** [review]

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java [hunk 7]
```diff
@@ -1494,6 +1495,30 @@ private void finishOptimize(ConnectorSession session, IcebergTableExecuteHandle
...
         commit(rewriteFiles, session);
         transaction.commitTransaction();
+
+        // TODO (https://github.com/trinodb/trino/issues/15439) this may not exactly be the snapshot we committed, if there is another writer
+        long newSnapshotId = transaction.table().currentSnapshot().snapshotId();
+        transaction = null;
+
+        // TODO (https://github.com/trinodb/trino/issues/15439): it would be good to publish data and stats atomically
+        beforeWriteSnapshotId.ifPresent(previous ->
+                verify(previous != newSnapshotId, "Failed to get new snapshot ID "));
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:1505]: The error message has a redundant following space. Could you remove or append the snapshot id?

You can also fix `finishInsert` method in a preparatory commit.

**Review by ebyhr — APPROVED** [review]

**Pull request merged**